html/captive-portal/lib/captiveportal/PacketFence/DynamicRouting/Module/Authentication/OAuth/Google.pm

```diff
@@ -13,8 +13,62 @@ Google OAuth module
 use Moose;
 extends 'captiveportal::DynamicRouting::Module::Authentication::OAuth';
 
+use pf::log;
+use pf::auth_log;
+
 has '+source' => (isa => 'pf::Authentication::Source::GoogleSource');
 
+=head2 handle_callback
+
+Google override to handle the callback from the OAuth2 provider and fetch the protected resource
+
+=cut
+
+sub handle_callback {
+    my ($self) = @_;
+
+    my $token = $self->get_token();
+    return unless($token);
+
+    # request a JSON response
+    my $h = HTTP::Headers->new( 'x-li-format' => 'json' );
+    my $response = $token->get($self->source->{'protected_resource_url'}, $h ); 
+
+    if ($response->is_success) {
+        my $info = $self->_decode_response($response); 
+        my $pid = $self->_extract_username_from_response($info);
+        my $hd = $info->{hd};
+        my $hosted_domain = $self->source->hosted_domain;
+        # get_logger->info(sub { use Data::Dumper; "OAuth2  response : ".Dumper($info) });   
+        if ( defined $hosted_domain && $hosted_domain ne "" && (!defined $hd || $hosted_domain ne $hd)){
```

Comment:
It would be better to modify the OAuth.pm authentication module so that it calls a filter_username function that would be overriden in this module to have this logic.

That way, this doesn't end up with duplicated code

Reply:
Sounds good. I'll try to refactor it in this way. Perl is far from my most proficient language thus why the copy-paste. I will try to refactor